File: oop/rich/src/main/java/de/hhu/rich/warenkorb/application/ApplicationService.java
package de.hhu.rich.warenkorb.application;

import de.hhu.rich.warenkorb.domain.model.artikel.Artikel;
import de.hhu.rich.warenkorb.domain.model.artikel.ArtikelID;
import de.hhu.rich.warenkorb.domain.model.common.anzahl.Anzahl;
import de.hhu.rich.warenkorb.domain.model.common.preis.Preis;
import de.hhu.rich.warenkorb.domain.model.kunde.KundeID;
import de.hhu.rich.warenkorb.domain.model.warenkorb.Warenkorb;
import de.hhu.rich.warenkorb.domain.model.warenkorb.WarenkorbID;
import de.hhu.rich.warenkorb.domain.model.warenkorb.WarenkorbRepository;
import de.hhu.rich.warenkorb.domain.model.artikel.ArtikelRepository;

import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.UUID;

@Service
public class ApplicationService {
    private final WarenkorbRepository warenkorbRepository;
    private final ArtikelRepository artikelRepository;


    public ApplicationService(WarenkorbRepository warenkorbRepository,
                              ArtikelRepository artikelRepository) {
        this.warenkorbRepository = warenkorbRepository;
        this.artikelRepository = artikelRepository;
    }

    public void erstelleWarenkorbFuerKunde(UUID warenkorbID, UUID kundeID, BigDecimal preis) {
        Preis maxEinkaufswert = new Preis(preis, "EUR");
        Warenkorb warenkorb = new Warenkorb(
            new WarenkorbID(warenkorbID),
            new KundeID(kundeID),
            maxEinkaufswert);
        warenkorbRepository.speichere(warenkorb);
    }

    public Warenkorb getWarenkorb(UUID warenkorb) {
        WarenkorbID warenkorbID = new WarenkorbID(warenkorb);
        return warenkorbRepository.findeMit(warenkorbID);
    }

    public Artikel getArtikel(UUID artikel) {
        ArtikelID artikelID = new ArtikelID(artikel);
        return artikelRepository.findeMit(artikelID);
    }

    public void fuegeArtikelZumWarenkorbHinzu(UUID artikelId, int anzahl, UUID warenkorbID) {
        Warenkorb warenkorb = getWarenkorb(warenkorbID);
        Artikel artikel = getArtikel(artikelId);

        warenkorb.fuegeHinzu(artikel, new Anzahl(anzahl));

        warenkorbRepository.speichere(warenkorb);
    }

    public void entferneArtikelAusWarenkorb(UUID artikelId, UUID warenkorbID) {
        Warenkorb warenkorb = getWarenkorb(warenkorbID);
        Artikel artikel = getArtikel(artikelId);

        warenkorb.entferne(artikel);

        warenkorbRepository.speichere(warenkorb);
    }


    public void reduziereAnzahlVonArtikelImWarenkorb(UUID artikelId, int anzahl, UUID warenkorbID) {
        Warenkorb warenkorb = getWarenkorb(warenkorbID);
        Artikel artikel = getArtikel(artikelId);

        warenkorb.reduziere(artikel, new Anzahl(anzahl));

        warenkorbRepository.speichere(warenkorb);
    }
}

File: fp/rich/src/main/clojure/de/hhu/rich/warenkorb/application/ApplicationService.clj
(ns de.hhu.rich.warenkorb.application.ApplicationService
  (:require [de.hhu.rich.warenkorb.domain.model.anzahl.Anzahl :as anzahl]
            [de.hhu.rich.warenkorb.domain.model.artikel.ArtikelRepository :as artikel-repository]
            [de.hhu.rich.warenkorb.domain.model.preis.Preis :as preis]
            [de.hhu.rich.warenkorb.domain.model.warenkorb.Warenkorb :as warenkorb]
            [de.hhu.rich.warenkorb.domain.model.warenkorb.WarenkorbRepository :as warenkorb-repository]))

(defn erstelle-warenkorb-fuer-kunde [kunde-id preis]
  (let [max-einkaufswert (preis/erstelle-preis preis "EUR")
        warenkorb (warenkorb/erstelle-warenkorb kunde-id max-einkaufswert)]
    (warenkorb-repository/add! warenkorb)))

(defn get-warenkorb [warenkorb-id]
  (warenkorb-repository/find warenkorb-id))

(defn get-artikel [artikel-id]
  (artikel-repository/find artikel-id))

(defn lege-artikel-in-den-warenkorb [artikel-id anzahl warenkorb-id]
  (let [warenkorb (get-warenkorb warenkorb-id)
        artikel (get-artikel artikel-id)]
    (warenkorb/fuege-hinzu warenkorb artikel (anzahl/erstelle-anzahl anzahl))
    (warenkorb-repository/update! warenkorb)))

(defn entferne-artikel-aus-dem-warenkorb [artikel-id warenkorb-id]
  (let [warenkorb (get-warenkorb warenkorb-id)
        artikel (get-artikel artikel-id)]
    (warenkorb/entferne warenkorb artikel)
    (warenkorb-repository/update! warenkorb)))

(defn reduziere-anzahl-von-artikel-in-dem-warenkorb [artikel-id anzahl warenkorb-id]
  (let [warenkorb (get-warenkorb warenkorb-id)
        artikel (get-artikel artikel-id)]
    (warenkorb/reduziere warenkorb artikel (anzahl/erstelle-anzahl anzahl))
    (warenkorb-repository/update! warenkorb)))

